nsqlookup/resolver.go

```diff
@@ -145,42 +213,26 @@ func (r *ConsulResolver) Resolve(ctx context.Context) (list []string, err error)
 	}
 
 	if res, err = t.RoundTrip(req); err != nil {
-		return
+		return err
 	}
 	defer res.Body.Close()
 
 	switch res.StatusCode {
 	case http.StatusOK:
 	case http.StatusNotFound:
-		return
+		return err
 	default:
-		err = fmt.Errorf("error looking up %s on consul agent at %s: %d %s", service, address, res.StatusCode, res.Status)
-		return
-	}
-
-	var results []struct {
-		Address        string
-		ServiceAddress string
-		ServicePort    int
-	}
-
-	if err = json.NewDecoder(res.Body).Decode(&results); err != nil {
-		return
+		err = fmt.Errorf("error looking up %s on consul agent at %s: %d %s", endpoint, address, res.StatusCode, res.Status)
+		return err
 	}
-	list = make([]string, 0, len(results))
 
-	for _, r := range results {
-		host := r.ServiceAddress
-		port := r.ServicePort
+	body, _ := ioutil.ReadAll(res.Body)
 
-		if len(host) == 0 {
-			host = r.Address
-		}
-
-		list = append(list, net.JoinHostPort(host, strconv.Itoa(port)))
+	if err = json.Unmarshal(body, result); err != nil {
```

Comment:
Instead of loading the response in a temporary buffer and then parsing it you could use a decoder to parse JSON from the response body direction:
```go
if err = json.NewDecoder(res.Body).Decode(result); err != nil {
   ...
}
```